libbeat/outputs/kafka/client.go

```diff
@@ -137,9 +142,16 @@ func (c *client) Publish(batch publisher.Batch) error {
 		msg.ref = ref
 		msg.initProducerMessage()
 		ch <- &msg.msg
+		select {
+		case publishError = <-ref.error:
+			ref.batch.RetryEvents(ref.failed)
+			break
+		case _ = <-ref.success:
+			// do nothing
+		}
```

Comment:
I'm afraid this one is much too expensive. The publish method receives a batch of events. This batch is converted into messages that can be consumed by the sarama lib and forwarded one by one. The lib will route the event to the correct brokers workers, where events are (after being load-balanced on the multiple brokers) will be batched up. The batch will eventually (after timeout or buffer full) send to kafka, and we get the response. We call the ref callbacks after the response from kafka. -> waiting on every single message will require us to have a full round-trip per event to be published.

The SyncProducer would be a better fit here, as this will support batches. But I'd prefer to keep the async client. E.g. on error we could set an error flag/value in the client. The next time `Publish` is called and the error flag is set, we backoff and promote the error to 'last-error'. If we do another send and get a new error from the next msg-ref, the we increase backoff. Due to the async nature of the sarama client, we might capture errors a little late, but at least we can still back-off.

The backoff client from the outputs package is no good fit for this. The backoff client will attempt to reconnect. But the sarama lib already handles reconnects internally. That's why we will need some custom logic in the `Publish` method only.